Document `code_paths` limitation when loading multiple models (#12471)
mlflow/mlflow: docs/source/model/dependencies.rst

Author: /ai hello, how would title this section?

    By the same reason, the ``code_paths`` option doesn't handle the relative import of ``code_paths=["../src"]``.

Limitation of ``code_paths`` in loading multiple models with the same module name but different implementations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The current implementation of the ``code_paths`` option has a limitation that it doesn't support reloading the model with the same module name but different implementations. The following example illustrates this limitation.

Reviewer: ```suggestion
The current implementation of the ``code_paths`` option has a limitation that it doesn't support simultaneously loading multiple models that depends on modules with the same name but different implementations. The following example illustrates this limitation.
```
nit: per my understanding, this might be more accurate description for the situation?

Author: Good point. More accurate way to describe this is `loading multiple models ... within the same python session`. `simultaneously` might make users wonder "what if they are loaded one by one"?

Reviewer: Makes sense!

        code_template = """
    import mlflow

    class MyModel(mlflow.pyfunc.PythonModel):
        def predict(self, context, model_input):
            return [{n}] * len(model_input)
    """

        my_model_path.write_text(code_template.format(n=1))

        sys.path.insert(0, str(tmpdir))
        import my_model

        # model 1
        model1 = my_model.MyModel()
        assert model1.predict(context=None, model_input=[0]) == [1]

        with mlflow.start_run():
            info1 = mlflow.pyfunc.log_model(
                artifact_path="model",
                python_model=model1,
                code_paths=[my_model_path],
            )

        # model 2
        my_model_path.write_text(code_template.format(n=2))
        importlib.reload(my_model)
        model2 = my_model.MyModel()
        assert model2.predict(context=None, model_input=[0]) == [2]

        with mlflow.start_run():
            info2 = mlflow.pyfunc.log_model(
                artifact_path="model",
                python_model=model2,
                code_paths=[my_model_path],
            )

        shutil.rmtree(tmpdir)
        sys.path.remove(str(tmpdir))

>         sys.modules.pop("my_model")

Reviewer: Can we add a comment for what these lines for?

Author: added!

        # Now we have two models with the same module name but different implementations.
        # Let's load them and check the prediction results.
        pred = mlflow.pyfunc.load_model(info1.model_uri).predict([0])
        assert pred == [1], pred  # passes

        # As the `my_model` module was loaded and cached in the previous `load_model` call,
        # the next `load_model` call will reuse it and return the wrong prediction result.
        assert "my_model" in sys.modules
        pred = mlflow.pyfunc.load_model(info2.model_uri).predict([0])
        assert pred == [2], pred  # doesn't pass, `pred` is [1]

To work around this limitation, you can remove the module from the cache before loading the model. For example:

.. code-block:: python

    model1 = mlflow.pyfunc.load_model(info1.model_uri).predict([0])
    sys.modules.pop("my_model")

>     model2 = mlflow.pyfunc.load_model(info1.model_uri).predict([0])

Reviewer: ```suggestion
    model2 = mlflow.pyfunc.load_model(info2.model_uri).predict([0])
```

Author: thanks for the catch!

Another workaround is to use different module names for different implementations. For example:

.. code-block:: python

    mlflow.pyfunc.log_model(
        artifact_path="model",

>         python_model=m1,

Author: ```suggestion
        python_model=mdoel1,
```
